File: docs-examples/example-java/src/test/java/io/micronaut/mqtt/docs/properties/ProductListener.java
package io.micronaut.mqtt.docs.properties;

// tag::imports[]
import io.micronaut.mqtt.annotation.Topic;
import io.micronaut.mqtt.annotation.MqttSubscriber;
import io.micronaut.mqtt.v5.annotation.MqttProperty;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
// end::imports[]

import io.micronaut.context.annotation.Requires;

@Requires(property = "spec.name", value = "PropertiesSpec")
// tag::clazz[]
@MqttSubscriber
public class ProductListener {

    List<String> messageProperties = Collections.synchronizedList(new ArrayList<>());

    @Topic("product")
    public void receive(byte[] data,
                        @MqttProperty("userId") String user,  // <2>
                        @Nullable @MqttProperty String contentType,  // <3>
                        @MqttProperty String appId) {  // <4>
        messageProperties.add(user + "|" + contentType + "|" + appId);
    }
}
// end::clazz[]

File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/mqtt/docs/properties/ProductListener.kt
package io.micronaut.mqtt.docs.properties

// tag::imports[]
import io.micronaut.mqtt.annotation.Topic
import io.micronaut.mqtt.annotation.MqttSubscriber
import io.micronaut.context.annotation.Requires
import io.micronaut.mqtt.v5.annotation.MqttProperty
import java.util.*
// end::imports[]

@Requires(property = "spec.name", value = "PropertiesSpec")
// tag::clazz[]
@MqttSubscriber
class ProductListener {

    val messageProperties: MutableList<String> = Collections.synchronizedList(ArrayList())

    @Topic("product")
    fun receive(data: ByteArray,
                @MqttProperty("userId") user: String, // <2>
                @MqttProperty contentType: String?, // <3>
                @MqttProperty appId: String) { // <4>
        messageProperties.add("$user|$contentType|$appId")
    }
}
// end::clazz[]
